### PR TITLE
cluster: refactor EIP712 domain data

### PR DESCRIPTION
**Pull request by xenowits**
Refactor `EIP712 Domain` data to add support for different chains ids.

category: refactor
ticket: https://github.com/ObolNetwork/charon/issues/1203

### CODE REVIEW

**Comment by xenowits** [comment]
Discussion: https://docs.google.com/document/d/1lYP1IU56JLtadrdaZTA9J1edGJjEaG8hLesyQlNxjCc/edit#heading=h.6ef9urxlm0ng

**Review by corverroos — CHANGES_REQUESTED** [review]
This breaks backwards compatibility with previous signatures. 

Also, you are not only changing the chainID, you are also changing name and salt. We said that the DV launchpad is going to sign like charon does, then we are going to get product to come up a proper next iteration. 

Lets rather hold off on this until we know what the plan is, how we are going to do backwards compatibility stuff, etc.

File: cluster/helpers.go [hunk 1]
```diff
@@ -165,14 +163,14 @@ func digestEIP712(address string, message []byte, nonce int) ([32]byte, error) {
 		},
 		PrimaryType: "Challenge",
 		Domain: apitypes.TypedDataDomain{
-			Name:    "ETHChallenger",
+			Name:    "Obol",
```
Comment (corverroos) [inline cluster/helpers.go:166]: remove these structure changes for now

File: cluster/cluster_internal_test.go [hunk 2]
```diff
@@ -28,21 +28,25 @@ import (
 )
 
 func TestDefinitionVerify(t *testing.T) {
+	forkVersion := "0x80000069"
```
Comment (corverroos) [inline cluster/cluster_internal_test.go:31]: make these constants in `helpers.go`

**Review by corverroos — APPROVED** [review]

File: cluster/helpers.go [hunk 3]
```diff
@@ -191,6 +174,68 @@ func digestEIP712(address string, message []byte, nonce int) ([32]byte, error) {
 	return crypto.Keccak256Hash(rawData), nil
 }
 
+// configHashSig returns the EIP712 (https://eips.ethereum.org/EIPS/eip-712) signature of the config hash.
+func configHashSig(secret *ecdsa.PrivateKey, configHash string, chainID int64) ([]byte, error) {
```
Comment (xenowits) [inline cluster/helpers.go:178]: will add tests for these new functions in the next PR

File: cluster/helpers.go [hunk 4]
```diff
@@ -191,6 +174,68 @@ func digestEIP712(address string, message []byte, nonce int) ([32]byte, error) {
...
+	if err != nil {
+		return nil, errors.Wrap(err, "sign EIP712")
+	}
+
+	return sig, nil
+}
+
+// configHashDigest returns the EIP712 digest of the config hash.
+func configHashDigest(configHash string, chainID int64) ([32]byte, error) {
+	signer := eip712SignerData(chainID)
```
Comment (corverroos) [inline cluster/helpers.go:194]: this isn't a `signer`, it is more `data`

File: cluster/helpers.go [hunk 5]
```diff
@@ -129,54 +132,34 @@ func aggSign(secrets [][]*bls_sig.SecretKeyShare, message []byte) ([]byte, error
...
-	}
-
-	return sig, nil
-}
-
-// digestEIP712 returns a EIP712 digest hash.
-// See reference https://medium.com/alpineintel/issuing-and-verifying-eip-712-challenges-with-go-32635ca78aaf.
-func digestEIP712(address string, message []byte, nonce int) ([32]byte, error) {
+// eip712SignerData returns an EIP712 structured data containing the signing domain and the required types.
+func eip712SignerData(chainID int64) apitypes.TypedData {
```
Comment (corverroos) [inline cluster/helpers.go:136]: Why not replace multiple functions with just a single function:
```
// digestEIP712 returns an EIP712 structured data containing the signing domain and the required types.
func digestEIP712(primaryType, fieldName, fieldValue string, chainID int64) ([]byte, error) {
    data := apitypes.TypedData{
        Types: apitypes.Types{
            "EIP712Domain": []apitypes.Type{
                {Name: "name", Type: "string"},
                {Name: "version", Type: "string"},
                {Name: "chainId", Type: "uint256"},
            },
            primaryType: []apitypes.Type{
                {Name: fieldName, Type: "string"},
            },
        },
        PrimaryType: primaryType,
        Message: apitypes.TypedDataMessage{
            fieldName: fieldValue,
        },
        Domain: apitypes.TypedDataDomain{
            Name:    "Obol",
            Version: "1",
            ChainId: ethmath.NewHexOrDecimal256(chainID),
        },
    }

    resp, _ , err := apitypes.TypedDataAndHash(data)

    return resp, err
}
```

**Review by corverroos — CHANGES_REQUESTED** [review]

File: cluster/helpers.go [hunk 6]
```diff
@@ -42,6 +42,13 @@ const (
 	k1SigLen = 65
 	// k1RecIdx is the secp256k1 signature recovery id index.
 	k1RecIdx = 64
+	// Fork version of sepolia testnet.
+	sepoliaForkVersion = "0x90000069"
```
Comment (corverroos) [inline cluster/helpers.go:46]: go comments should have the variable name as the prefix: `// sepoliaForkVersion is the Sepolia testnet fork version. `

Comment (corverroos) [reply cluster/helpers.go:46]: move this to a test file if only used in testing.

File: cluster/helpers_internal_test.go [hunk 7]
```diff
@@ -68,3 +68,16 @@ func TestVerifySig(t *testing.T) {
 		require.True(t, ok)
 	})
 }
+
+func TestForkVersionToChainID(t *testing.T) {
+	gnosisForkVersion := []byte{0, 0, 0, 100}
```
Comment (corverroos) [inline cluster/helpers_internal_test.go:73]: use sepolia or make this a constant?

File: cluster/helpers.go [hunk 8]
```diff
@@ -263,3 +256,21 @@ func from0xHex(s string, length int) ([]byte, error) {
...
+	switch fmt.Sprintf("%#x", forkVersion) {
+	case "0x00000000": // Mainnet
+		return 1, nil
+	case "0x00001020": // Goerli/Prater
+		return 5, nil
+	case "0x00000064": // Gnosis Chain
+		return 100, nil
+	case "0x80000069": // Ropsten
+		return 3, nil
+	case "0x90000069": // Sepolia
```
Comment (corverroos) [inline cluster/helpers.go:271]: I suggest make these all constants

**Review by corverroos — APPROVED** [review]

File: cluster/helpers.go [hunk 9]
```diff
@@ -42,6 +42,13 @@ const (
 	k1SigLen = 65
 	// k1RecIdx is the secp256k1 signature recovery id index.
 	k1RecIdx = 64
+	// Fork version of sepolia testnet.
+	sepoliaForkVersion = "0x90000069"
+	// EIP712 values.
+	configHashPrimaryType = "ConfigHash"
```
Comment (corverroos) [inline cluster/helpers.go:48]: suggest identical prefixes:
```
eip712TypeConfigHash = "ConfigHash"
eip712FieldConfigHash = "config_hash"
eip712TypeENR = "ENR"
eip712FieldENR = "enr"
```

Comment (corverroos) [reply cluster/helpers.go:48]: another option would be to make this a type:

```
type eip712Type struct {
  PrimaryType string
  Field string
}

var (
  eip712TypeConfigHash = eip712Type {"ConfigHash","config_hash}
  eip712TypeENR = eip712Type {"ENR","enr}
)

func signEIP712(secret *ecdsa.PrivateKey, typ eip712Type, fieldValue string, chainID int64) ([]byte, error) {
```

File: cluster/cluster_test.go [hunk 10]
```diff
@@ -44,7 +44,7 @@ func TestEncode(t *testing.T) {
 				3,
 				testutil.RandomETHAddress(),
 				testutil.RandomETHAddress(),
-				"0x00000002",
+				"0x90000069", // Sepolia testnet.
```
Comment (corverroos) [inline cluster/cluster_test.go:47]: use the constant

Comment (xenowits) [reply cluster/cluster_test.go:47]: will raise a PR to combine network chainID, name and forkVersion and replace relevant constants across the codebase

**Pull request merged**